Build summary table without innerHTML

otherdaniel opened the pull request:
The summary table is currently built up as a (potentially rather long) text string and then passed to `innerHTML`. Let's instead use DOM methods to build it up.

This fixes the issue in https://github.com/web-platform-tests/wpt/issues/44352. It might also improve readability.

Results generation was measured for the test with the largest result table (html/dom/reflection-text.text), which was around 800ms. The comment warning about "tens of seconds of execution time" from textContent calls alone seems to no longer be accurate.

Ms2ger commented on resources/testharness.js:
```diff
@@ -4296,81 +4287,84 @@
-                    "</pre></td></tr>";
+                row_other_info.textContent = output_fn;
+
+                table.appendChild(row.cloneNode(true));
```
Is there a reason behind using `cloneNode` here? If so, please document.

Ms2ger replied:
Hm, I don't see it. You're moving some element creation into the browser implementation (I'd guess string copies under this would be refcounted), but I don't think you're actually saving any `textContent` sets. The actual perf impact is anyone's guess, of course. If you've got time to compare, I think using `render` directly would be more readable, at least.

otherdaniel replied:
Updated the PR to use `render()`. It certainly is more readable.

Ms2ger commented on resources/testharness.js:
```diff
@@ -4254,18 +4257,6 @@
                              });
                 });
 
-        // This use of innerHTML plus manual escaping is not recommended in
-        // general, but is necessary here for performance.  Using textContent
-        // on each individual <td> adds tens of seconds of execution time for
-        // large test suites (tens of thousands of tests).
```
I assume you've tested this? Please explain in the PR summary

otherdaniel replied:
On my machine; current version of Chrome; on the largest result table that we found (html/dom/reflection-text.text; with indeed >10k entries) the results generation took ~800ms in total. So the total time taken by the `textContent` calls would be clearly less than "tens of seconds".

Ms2ger replied:
Could you also compare in Firefox and Safari?

otherdaniel replied:
On Chrome, I measured (with the latest version) also about ~800ms for the results generation, vs. ca 650ms for the original code.
On Firefox, I get about 1100ms, vs ~800ms previously.
I don't have a machine that Safari runs on.

I'll note that both of those measurements have very high variance, and on some runs the old code was just as slow as the new one; but usually it seemed a little faster. I'm not seeing the "tens of seconds of execution time" that the comment promises, anywhere.

How is this usually measured? Is there a way to get performance data out of the bot runs that are done for each PR?

Ms2ger reviewed (approved):
Looks good overall, but I'd like James to approve as well.

otherdaniel commented on resources/testharness.js:
```diff
@@ -4296,81 +4287,62 @@
-                var prefix_len = output_fn.length;
-                var output_args = assert.args;
-                var output_len = output_args.reduce((prev, current) => prev+current, prefix_len);
-                if (output_len[output_len.length - 1] > 50) {
-                    output_args = output_args.map((x, i) =>
-                    (i > 0 ? "  ".repeat(prefix_len) : "" )+ x + (i < output_args.length - 1 ? ",\n" : ""));
-                } else {
-                    output_args = output_args.map((x, i) => x + (i < output_args.length - 1 ? ", " : ""));
-                }
```
I dropped that on purpose, since this appears to be dead code:

* assert_args is an array with the arguments passed to the original `AssertRecord`, formatted by `format_value`.
* The last argument seems to always be a string.
* Since `format_value` with put strings in quotes, the last character of the last string is always `"`.
* Reducing this over `+` will produce a string whose last character is `"`.
* Comparing that last character against 50 will always be `false`.

Example:
- `assert_equals(2, 3, "explainy explainy")`
- => `[2,3, '"explainy explainy"' ]`
- => '23"explainy explainy"'

I *think* that is meant to do line breaking. But I don't think it actually does this.

I tried to rewrite this to do line breaking. But firstly, that'd be a change from what we had, and secondly, that looked worse than just letting the browser do the line breaking. 

Maybe I'm misunderstanding this; but in that case I will need a bit of a hint of where I went wrong.

Original discussion here: https://chromium-review.googlesource.com/c/chromium/src/+/5290877/comment/b3ec4e27_c7bdddba/

Ms2ger replied:
I don't think "The last argument seems to always be a string." is correct, but indeed `output_len` is a string and I'm pretty sure there's no single character that can be `> 50`. I imagine
```js
var output_len = output_args.reduce((prev, current) => prev+current.length, prefix_len);
if (output_len > 50) {
```
was meant, to achieve an output like
```js
assert_equals("looooooooooooooooooooooooooooooong string",
              "looooooooooooooooooooooooooooooong string")
```
However, it doesn't indent the following lines by the length of `assert_equals(` but the length of `<strong>assert_equals</strong>(`, which means that the rendered code doesn't line up anyway.

Since nobody noticed or cared enough to figure out what was wrong with this code in the three years since it was added in #27224, I think you're justified in removing it.

otherdaniel commented on resources/testharness.js:
```diff
@@ -4296,81 +4287,62 @@
             var asserts = asserts_run_by_test.get(test);
             if (!asserts) {
-                return "No asserts ran";
+                asserts_output.querySelector("summary").textContent = "No asserts ran";
```
Yes, you're right. It now looks like the original code.

Done.

jgraham commented:
In gecko the profiler shows a 85ms (~22%) regression in the performance of `Output.prototype.show_results` in `/html/dom/reflection-text.html` in debug mode (so we also get the assert output).

It seems like browsers are now enough faster that we can probably remove the optimistation.

otherdaniel commented on resources/testharness.js:
```diff
@@ -4296,81 +4287,62 @@
         function get_asserts_output(test) {
+            const asserts_output = render(
+                ["details", {},
+                    ["summary", {}, ""],
```
Done.

otherdaniel committed:
git squash commit for wpt-summary-table.

fa98516f965cf612122f9848c4189f326c435d6d
Build summary table without innerHTML

aefe44d96878ecd8b34d56a593027d800607591d
Replace node creation & cloning with render() calls.

376641ecbb48d5097044f0f4c0b848d7a426ac6f
Review feedback
